crates/typst/src/layout/grid/layout.rs

```diff
@@ -411,204 +429,311 @@ impl CellGrid {
+        // least 'children.len()' cells (if no explicit lines were specified),
         // even though some of them might be placed in arbitrary positions and
         // thus cause the grid to expand.
         // Additionally, make sure we allocate up to the next multiple of 'c',
         // since each row will have 'c' cells, even if the last few cells
         // weren't explicitly specified by the user.
         // We apply '% c' twice so that the amount of cells potentially missing
-        // is zero when 'items.len()' is already a multiple of 'c' (thus
-        // 'items.len() % c' would be zero).
-        let items = items.into_iter();
-        let Some(item_count) = items.len().checked_add((c - items.len() % c) % c) else {
+        // is zero when 'children.len()' is already a multiple of 'c' (thus
+        // 'children.len() % c' would be zero).
+        let children = children.into_iter();
+        let Some(child_count) = children.len().checked_add((c - children.len() % c) % c)
+        else {
             bail!(span, "too many cells or lines were given")
         };
-        let mut resolved_cells: Vec<Option<Entry>> = Vec::with_capacity(item_count);
-        for item in items {
-            let cell = match item {
-                GridItem::HLine { y, start, end, stroke, span, position } => {
-                    let y = y.unwrap_or_else(|| {
-                        // When no 'y' is specified for the hline, we place it
-                        // under the latest automatically positioned cell.
-                        // The current value of the auto index is always the
-                        // index of the latest automatically positioned cell
-                        // placed plus one (that's what we do in
-                        // 'resolve_cell_position'), so we subtract 1 to get
-                        // that cell's index, and place the hline below its
-                        // row. The exception is when the auto_index is 0,
-                        // meaning no automatically positioned cell was placed
-                        // yet. In that case, we place the hline at the top of
-                        // the table.
-                        auto_index
-                            .checked_sub(1)
-                            .map_or(0, |last_auto_index| last_auto_index / c + 1)
-                    });
-                    if end.is_some_and(|end| end.get() < start) {
-                        bail!(span, "line cannot end before it starts");
+        let mut resolved_cells: Vec<Option<Entry>> = Vec::with_capacity(child_count);
+        for child in children {
+            let mut is_header = false;
+            let mut header_start = usize::MAX;
+            let mut header_end = 0;
+            let mut header_span = Span::detached();
+            let mut min_auto_index = 0;
+
+            let (header_items, simple_item) = match child {
+                ResolvableGridChild::Header { repeat, span, items, .. } => {
```

Comment:
would it make sense to turn this children -> items conversion into a separate function returning an iterator? That way we can skip one level of indent and for loop nesting and it gets clearer that this isn't some O(n^2) operation. it would also simplify the diff a lot.

Reply:
You mean using something like `flat_map`? I thought about it, but it feels like it would be less clean to handle the post-items logic and the per-child state (the `header_` variables above). The state would then appear as part of the iterator, e.g. some sort of "items left to end header" or something. Not really sure what's the best approach here.

Reply:
Yeah I see. It's very challenging to keep the complexity at bay here.

Reply:
By the way, the diff becomes much simpler if you disable whitespace preview (at the :gear: at the top of the diff in the GitHub website). Though I guess not so much in Git itself. :p